src/main/java/org/apache/xml/security/encryption/XMLCipher.java

```diff
@@ -1500,42 +1542,77 @@ public Key decryptKey(EncryptedKey encryptedKey, String algorithm)
-                    mgfParameterSpec = new MGF1ParameterSpec("SHA-224");
-                } else if (EncryptionConstants.MGF1_SHA256.equals(mgfAlgorithm)) {
-                    mgfParameterSpec = new MGF1ParameterSpec("SHA-256");
-                } else if (EncryptionConstants.MGF1_SHA384.equals(mgfAlgorithm)) {
-                    mgfParameterSpec = new MGF1ParameterSpec("SHA-384");
-                } else if (EncryptionConstants.MGF1_SHA512.equals(mgfAlgorithm)) {
-                    mgfParameterSpec = new MGF1ParameterSpec("SHA-512");
-                }
-            }
-            return new OAEPParameterSpec(jceDigestAlgorithm, "MGF1", mgfParameterSpec, pSource);
+        if (keyAgreementParameter.getOriginatorPrivateKey() == null) {
+            LOG.log(Level.DEBUG, "Originator's private key is not set in keyAgreementParameter, " +
+                    "generate an ephemeral key for the originator's private key.");
+            KeyPair originatorKeyPair = KeyUtils.generateEphemeralDHKeyPair(
+                    keyAgreementParameter.getRecipientPublicKey(), null);
+            keyAgreementParameter.setOriginatorKeyPair(originatorKeyPair);
         }
+    }
 
+    /**
+     *  Method resolves the EncryptedKey  using the EncryptedKeyResolver
+     *
+     * @param encryptedKey
+     * @return
+     * @throws XMLEncryptionException
+     */
+    private AlgorithmParameterSpec getAlgorithmParameters(EncryptedKey encryptedKey) throws XMLSecurityException {
+        if (encryptedKey == null
+            || encryptedKey.getEncryptionMethod() == null
+            || encryptedKey.getEncryptionMethod().getAlgorithm() == null) {
+            LOG.log(Level.DEBUG,"EncryptedKey key algorithm is null");
+            return null;
+        }
+
+        EncryptionMethod encMethod = encryptedKey.getEncryptionMethod();
+        String encryptionAlgorithm = encMethod.getAlgorithm();
+
+        if (XMLCipher.RSA_OAEP.equals(encryptionAlgorithm)
+                || XMLCipher.RSA_OAEP_11.equals(encryptionAlgorithm)) {
+            LOG.log(Level.DEBUG,"EncryptedKey key algorithm is RSA OAEP");
+            return  XMLCipherUtil.constructOAEPParameters(
+                    encryptionAlgorithm, encMethod.getDigestAlgorithm(),
+                    encMethod.getMGFAlgorithm(), encMethod.getOAEPparams());
+        }
+
+        KeyInfoEnc keyInfo = encryptedKey.getKeyInfo() instanceof KeyInfoEnc ? (KeyInfoEnc) encryptedKey.getKeyInfo(): null;
+        if (keyInfo != null && keyInfo.containsAgreementMethod()) {
+            // resolve the agreement method
+            LOG.log(Level.DEBUG,"EncryptedKey key is using Key agreement data");
+            AgreementMethod agreementMethod = keyInfo.itemAgreementMethod(0);
+            return XMLCipherUtil.constructRecipientKeyAgreementParameters(encryptionAlgorithm,
+                    agreementMethod, (PrivateKey) this.key);
+        }
         return null;
```

Comment:
Should you be throwing an exception instead of returning null if the algorithm or parameters are not recognized or supported?

Reply:
I don't think an error should be thrown there because it would change the existing behavior of the code.

The method getAlgorithmParameters 
is a generic substitute of the constructOAEPParameters
see the: https://github.com/apache/santuario-xml-security-java/blob/main/src/main/java/org/apache/xml/security/encryption/XMLCipher.java#L1373
where I added the option to construct the KeyAgreementParameters.

Throwing an error would change the behavior of the method 
XMLCipher.decryptKey(EncryptedKey encryptedKey, String algorithm)
and would break decryption for Encryption algorithms where additional parameters are not needed.
In the case of encryption  algorithms where parameters are mandatory, the relevant exceptions are thrown in the particular method that constructs the parameters, as example in the:
 XMLCipherUtil.constructRecipientKeyAgreementParameters
 
 Regarding if the algorithm is not recognized or supported :
 A basic validation is already performed at cipher configuration phase 
  XMLCipher.constructCipher
  within the methods: 
  
  String jceAlgorithm = JCEMapper.translateURItoJCEID(algorithm)
  ...
  c = Cipher.getInstance(jceAlgorithm)